gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,476 @@
+
+      default:
+        throw new UnsupportedOperationException(fieldType + " is unsupported");
+    }
+  }
+
+  /**
+   * Converts a JsonElement into a supported ParquetType
+   * @author tilakpatidar
+   *
+   */
+  public static abstract class JsonElementConverter {
+    protected final JsonSchema jsonSchema;
+
+    protected JsonElementConverter(JsonSchema schema) {
+      this.jsonSchema = schema;
+    }
+
+    /**
+     * Convert value to a parquet type and perform null check.
+     * @param value
+     * @return Parquet safe type
+     */
+    public Object convert(JsonElement value) {
+      if (value.isJsonNull()) {
+        if (this.jsonSchema.isNullable()) {
+          return null;
+        }
+        throw new RuntimeException(
+            "Field: " + this.jsonSchema.getColumnName() + " is not nullable and contains a null value");
+      }
+      return convertField(value);
+    }
+
+    /**
+     * Returns a {@link Type} parquet schema
+     * @return
+     */
+    abstract Type schema();
+
+    /**
+     * Convert JsonElement to Parquet type
+     * @param value
+     * @return
+     */
+    abstract Object convertField(JsonElement value);
+  }
+
+  /**
+   * Converts a {@link JsonSchema} to a {@link PrimitiveType}
+   */
+  public static abstract class PrimitiveConverter extends JsonElementConverter {
+    protected final boolean repeated;
+    private PrimitiveTypeName outputType;
+
+    /**
+     * @param jsonSchema
+     * @param repeated
+     * @param outputType
+     */
+    public PrimitiveConverter(JsonSchema jsonSchema, boolean repeated, PrimitiveTypeName outputType) {
+      super(jsonSchema);
+      this.repeated = repeated;
+      this.outputType = outputType;
+    }
+
+    protected Type schema() {
+      return new PrimitiveType(this.repeated ? REPEATED : this.jsonSchema.optionalOrRequired(), this.outputType,
+          this.jsonSchema.getColumnName());
+    }
+  }
+
+  /**
+   * Converts {@link JsonSchema} having collection of elements of {@link InputType} into a {@link GroupType}.
+   */
+  public static abstract class CollectionConverter extends JsonElementConverter {
+    private InputType elementType;
+    private JsonElementConverter elementConverter;
+
+    public CollectionConverter(JsonSchema collectionSchema, InputType elementType) {
+      super(collectionSchema);
+      this.elementType = elementType;
+    }
+
+    /**
+     * Prepare a {@link JsonSchema} for the elements in a collection.
+     * @param fieldName
+     * @return
+     */
+    protected JsonSchema getElementSchema(String fieldName) {
+      JsonSchema jsonSchema = JsonSchema.buildBaseSchema(this.elementType);
+      jsonSchema.setColumnName(fieldName);
+      return jsonSchema;
+    }
+
+    /**
+     * Set a {@link JsonElementConverter} for the elements within a collection.
+     * @param elementConverter
+     */
+    protected void setElementConverter(JsonElementConverter elementConverter) {
```

Comment:
If the `elementConverter` is not supposed to change once set, can it be passed as a constructor argument? Then we can remove method `setElementConverter`. It's perfectly find to make it `protected`, then we can get rid of `getElementConverter`.